src/trace.rs

```diff
@@ -0,0 +1,136 @@
+        tracing::debug!(
+            target: COMMAND_TRACING_EVENT_TARGET,
+            command = self.serialize_command_or_reply(event.command).as_str(),
+            database_name = event.db.as_str(),
+            command_name = event.command_name.as_str(),
+            request_id = event.request_id,
+            driver_connection_id = event.connection.id,
+            server_host = event.connection.address.host(),
+            server_port = event.connection.address.port(),
+            service_id = event
+                .service_id
+                .map(|id| id.tracing_representation())
+                .as_deref(),
+            "Command started"
+        );
+    }
+
+    fn handle_command_succeeded_event(&self, event: CommandSucceededEvent) {
+        tracing::debug!(
+            target: COMMAND_TRACING_EVENT_TARGET,
+            reply = self.serialize_command_or_reply(event.reply).as_str(),
+            command_name = event.command_name.as_str(),
+            request_id = event.request_id,
+            driver_connection_id = event.connection.id,
+            server_host = event.connection.address.host(),
+            server_port = event.connection.address.port(),
+            service_id = event
+                .service_id
+                .map(|id| id.tracing_representation())
+                .as_deref(),
+            duration_ms = event.duration.as_millis().tracing_representation().as_str(),
+            "Command succeeded"
+        );
+    }
+
+    fn handle_command_failed_event(&self, event: CommandFailedEvent) {
+        tracing::debug!(
+            target: COMMAND_TRACING_EVENT_TARGET,
+            failure = event.failure.tracing_representation().as_str(),
+            command_name = event.command_name.as_str(),
+            request_id = event.request_id,
+            driver_connection_id = event.connection.id,
+            server_host = event.connection.address.host(),
+            server_port = event.connection.address.port(),
+            service_id = event
+                .service_id
+                .map(|id| id.tracing_representation())
+                .as_deref(),
+            duration_ms = event.duration.as_millis().tracing_representation().as_str(),
+            "Command failed"
+        );
+    }
+}
+
+pub(crate) const COMMAND_TRACING_EVENT_TARGET: &str = "mongodb::command";
+
+trait TracingRepresentation {
+    fn tracing_representation(self) -> String;
+}
+
+impl TracingRepresentation for bson::oid::ObjectId {
+    fn tracing_representation(self) -> String {
+        self.to_hex()
+    }
+}
+
+impl TracingRepresentation for u128 {
+    fn tracing_representation(self) -> String {
+        self.to_string()
+    }
+}
+
+impl TracingRepresentation for bson::Document {
+    fn tracing_representation(self) -> String {
+        Bson::Document(self).into_canonical_extjson().to_string()
+    }
+}
+
+impl TracingRepresentation for crate::error::Error {
+    fn tracing_representation(self) -> String {
+        self.to_string()
```

Comment:
probably want to do something nicer here (or update the `Display` implementation to) but I haven't looked at it much yet

Reply:
Yeah, for server errors we could just want the extjson form. For the more basic ones we can just do the message. For structured rust ones, I'm not totally sure what to do, though the `Display` implementation will be at least decent for those too.